File: typescript/problem_5.ts
type Pair = (f: (a: number, b: number) => number) => number;

// cons constructs a pair encoded as a closure. car and cdr extract the first
// and second elements by passing selector functions into the pair.
export function cons(a: number, b: number): Pair {
    return (f) => f(a, b);
}

// car returns the first element of a pair.
export function car(pair: Pair): number {
    return pair((a, _) => a);
}

// cdr returns the second element of a pair.
export function cdr(pair: Pair): number {
    return pair((_, b) => b);
}

export function problem5(): void {
    console.log("Problem 5:");
    if (car(cons(3, 4))  !== 3)  throw new Error("assertion failed");
    if (cdr(cons(3, 4))  !== 4)  throw new Error("assertion failed");
    if (car(cons(0, -1)) !== 0)  throw new Error("assertion failed");
    if (cdr(cons(0, -1)) !== -1) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_5.rs
// cons constructs a pair encoded as a closure. car and cdr extract the first
// and second elements by passing selector functions into the pair.
pub fn cons(a: i64, b: i64) -> impl Fn(fn(i64, i64) -> i64) -> i64 {
    move |f| f(a, b)
}

// car returns the first element of a pair.
pub fn car(pair: impl Fn(fn(i64, i64) -> i64) -> i64) -> i64 {
    pair(|a, _| a)
}

// cdr returns the second element of a pair.
pub fn cdr(pair: impl Fn(fn(i64, i64) -> i64) -> i64) -> i64 {
    pair(|_, b| b)
}

pub fn problem5() {
    println!("Problem 5:");
    assert_eq!(car(cons(3, 4)),   3,  "assertion failed");
    assert_eq!(cdr(cons(3, 4)),   4,  "assertion failed");
    assert_eq!(car(cons(0, -1)),  0,  "assertion failed");
    assert_eq!(cdr(cons(0, -1)), -1,  "assertion failed");
    println!("All Tests Passed");
}
